### ISSUE TITLE
TASK: SOTA Research Initiative — Complete Methodology & Research Instructions

### ISSUE THREAD

**Issue by carochacs** [get-flashbacks/feedBack#13]
## Overview
This is the master methodology document for the FeedBack ecosystem SOTA (State-of-the-Art) tool review initiative. It provides researchers with step-by-step instructions for evaluating tool alternatives across 5 high-priority research areas.

**Related Research Issues:**
- [feedback-plugin-lyrics-karaoke#2](https://github.com/get-flashbacks/feedback-plugin-lyrics-karaoke/issues/2) — CREPE pitch extraction
- [feedback#4](https://github.com/get-flashbacks/feedback/issues/4) — demucs v4 & GP8 support
- [feedback-plugin-lyrics-sync#2](https://github.com/get-flashbacks/feedback-plugin-lyrics-sync/issues/2) — Faster-Whisper evaluation
- [feedback-plugin-chord-detector#3](https://github.com/get-flashbacks/feedback-plugin-chord-detector/issues/3) — Chord detection approaches
- [feedback-plugin-difficulty-ladder#31](https://github.com/get-flashbacks/feedback-plugin-difficulty-ladder/issues/31) — Kalman filter for adaptive difficulty

---

### Phase 1: Literature Review (2–4 hours per research area)

**Objective:** Understand the landscape of competing tools, academic context, and community feedback.

**Deliverables:**
- [ ] Identify 3–5 peer-reviewed papers or authoritative resources on the topic
- [ ] List GitHub repositories for each alternative tool
- [ ] Note any published benchmarks or comparison studies
- [ ] Collect community feedback (GitHub issues, discussions, Reddit, academic forums)

**Output format:**
```markdown
## Literature Review

### Papers
- [Paper Title](url) — [Key finding: X]
- [Paper Title](url) — [Key finding: Y]

### GitHub Repositories
- [Tool A](url) — [Status: active/archived], [Last update: date]
- [Tool B](url) — [Status: active/archived], [Last update: date]

### Benchmarks & Comparisons
- [Benchmark Name](url) — [Results summary]

### Community Feedback
- [Issue/Discussion](url) — [Key insight]
```

---

### Phase 2: Proof-of-Concept Setup (4–8 hours)

**Objective:** Get the alternative tool(s) running on test data to assess feasibility.

**Deliverables:**
- [ ] Install/configure each alternative alongside the current tool
- [ ] Prepare 3–5 test cases (sample audio, charts, or data)
- [ ] Run both implementations on the same test data
- [ ] Document setup steps and any blockers encountered
- [ ] Collect basic qualitative observations (speed, ease of use, quality)

**Output format:**
```python
# Example: Pitch extraction PoC
import librosa
import torchcrepe

audio, sr = librosa.load('test_vocal.wav')

# Current approach (pYIN)
f0_pyin, voiced_pyin, voicing_prob_pyin = librosa.pyin(
    audio, fmin=80, fmax=400, sr=sr
)

# Alternative (CREPE)
f0_crepe = torchcrepe.predict(
    audio, sr, viterbi=True, return_confidence=False
)

# Compare outputs
print(f"pYIN octave errors: {count_octave_errors(f0_pyin)}")
print(f"CREPE octave errors: {count_octave_errors(f0_crepe)}")
```

**Qualitative Notes:**
- Latency observations (inference time per sample)
- Memory usage / resource requirements
- Ease of integration into existing pipeline
- Any unexpected behavior or edge cases

---

### Phase 3: Benchmark & Evaluation (8–16 hours)

**Objective:** Quantitatively measure performance differences on a realistic dataset.

**Deliverables:**
- [ ] Select or create a benchmark dataset (see dataset guidance below)
- [ ] Define 2–3 primary evaluation metrics and success criteria
- [ ] Run both implementations on the full dataset
- [ ] Collect raw measurements (latency, accuracy, resource usage)
- [ ] Generate comparison visualizations (plots, tables, traces)
- [ ] Manual inspection / listening test (where applicable)

**Benchmark Datasets (by research area):**

| Research Area | Benchmark | Size | URL | Notes |
|---|---|---|---|---|
| **Pitch extraction** | MIR-1K | 1000 clips | http://www.music.cs.cmu.edu/mirdata/ | Chinese singing, vocal-heavy |
| **Pitch extraction** | VoxCeleb | 2000+ speakers | https://www.robots.ox.ac.uk/~vgg/data/voxceleb/ | Diverse vocals, languages |
| **Stem separation** | MUSDB18 | 150 songs | https://sigsep.github.io/musdb/ | Standard benchmark, 4 stems |
| **Chord detection** | Isophonics | 200+ songs | http://isophonics.cs.upf.edu/ | Annotated chord boundaries + symbols |
| **Lyrics alignment** | Common Voice | 1000+ clips | https://commonvoice.mozilla.org/ | Multi-language, singing subset available |

**Metrics by Research Area:**

**Pitch Extraction:**
```python
def octave_error_rate(predicted, ground_truth, threshold_cents=50):
    """% of frames with octave error (±1200 cents)"""
    errors = np.abs(predicted - ground_truth)
    octave_errors = (np.abs(errors - 1200) < threshold_cents) | \
                    (np.abs(errors + 1200) < threshold_cents)
    return np.mean(octave_errors) * 100

def voicing_accuracy(predicted_voiced, ground_truth_voiced):
    """% frames where voicing detection agrees"""
    return np.mean(predicted_voiced == ground_truth_voiced) * 100

def raw_pitch_accuracy(predicted, ground_truth, threshold_cents=50):
    """% frames within threshold_cents of ground truth"""
    errors = np.abs(predicted - ground_truth)
    return np.mean(errors < threshold_cents) * 100
```

**Stem Separation:**
```python
def signal_distortion_ratio(reference, predicted):
    """Higher is better (0-inf, typical range 5-15 dB)"""
    s_target = np.inner(predicted, reference) / np.inner(reference, reference) * reference
    e_noise = predicted - s_target
    return 10 * np.log10(np.sum(s_target**2) / (np.sum(e_noise**2) + 1e-8))

# Run on MUSDB18 subset:
for track in musdb.Dataset(split='test'):
    estimates = separate(track.audio)  # Both v3 and v4
    for source in ['vocals', 'drums', 'bass', 'other']:
        sdr_v3 = signal_distortion_ratio(track.sources[source], estimates_v3[source])
        sdr_v4 = signal_distortion_ratio(track.sources[source], estimates_v4[source])
        print(f"{track.name} {source}: v3={sdr_v3:.2f} dB, v4={sdr_v4:.2f} dB")
```

**Chord Detection:**
```python
def chord_accuracy(predicted_chords, ground_truth_chords, frame_rate=10):
    """% frames where predicted chord symbol matches ground truth"""
    # Chords are time-stamped; align to frame rate
    matches = sum(1 for pred, gt in zip(predicted_chords, ground_truth_chords) if pred == gt)
    return matches / len(ground_truth_chords) * 100

def boundary_accuracy(predicted_boundaries, ground_truth_boundaries, tolerance_sec=0.5):
    """% chord boundaries within tolerance_sec"""
    matches = 0
    for pred_t in predicted_boundaries:
        if any(abs(pred_t - gt_t) < tolerance_sec for gt_t in ground_truth_boundaries):
            matches += 1
    return matches / len(ground_truth_boundaries) * 100 if ground_truth_boundaries else 0
```

**Lyrics Alignment (Whisper variants):**
```python
def word_error_rate(predicted_transcript, ground_truth_transcript):
    """Lower is better (0-100+). Standard NLP metric."""
    ref_words = ground_truth_transcript.split()
    pred_words = predicted_transcript.split()
    # Edit distance / reference length
    from difflib import SequenceMatcher
    ratio = SequenceMatcher(None, pred_words, ref_words).ratio()
    return (1 - ratio) * 100

def alignment_accuracy(predicted_times, ground_truth_times, tolerance_ms=50):
    """% syllables with timing within tolerance_ms"""
    matches = sum(1 for pt, gt in zip(predicted_times, ground_truth_times) 
                  if abs((pt - gt) * 1000) < tolerance_ms)
    return matches / len(ground_truth_times) * 100
```

**Adaptive Difficulty (Kalman vs. EMA):**
```python
def engagement_retention(player_sessions_algorithm_a, player_sessions_algorithm_b):
    """% of players who complete N songs"""
    complete_a = sum(1 for s in player_sessions_algorithm_a if s.completed_songs >= 3)
    complete_b = sum(1 for s in player_sessions_algorithm_b if s.completed_songs >= 3)
    return {
        'algorithm_a': complete_a / len(player_sessions_algorithm_a),
        'algorithm_b': complete_b / len(player_sessions_algorithm_b),
        'improvement': (complete_b - complete_a) / complete_a * 100
    }

def difficulty_stability(mastery_adjustments_per_song):
    """Lower variance = more stable algorithm"""
    return np.std(mastery_adjustments_per_song)
```

**Output format (CSV for comparison):**
```
Song,Tool,Metric,Value,Latency_ms,CPU%
test_vocal_1.wav,pYIN,octave_error_rate,2.1,120,15
test_vocal_1.wav,CREPE,octave_error_rate,1.3,450,45
test_vocal_2.wav,pYIN,octave_error_rate,5.4,120,15
test_vocal_2.wav,CREPE,octave_error_rate,0.9,450,45
...
```

**Visualizations:**
- [ ] Line plot: metric vs. test case
- [ ] Box plot: distribution comparison (mean, median, quartiles)
- [ ] Scatter: accuracy vs. latency trade-off
- [ ] Waterfall chart: relative improvement (%) across test set

---

### Phase 4: Decision & Report (2–4 hours)

**Objective:** Synthesize findings and make adoption recommendation.

**Deliverables:**
- [ ] Summary of Phase 1–3 findings
- [ ] Clear **adoption decision** with reasoning
- [ ] Implementation plan (if adopting alternative)
- [ ] Risk assessment and migration cost estimate
- [ ] Next steps and owner assignment

**Decision Framework:**

```
IF (primary_metric_improvement > threshold) AND (latency_acceptable) AND (integration_effort_low):
    RECOMMENDATION = "ADOPT"
ELIF (primary_metric_improvement > threshold) BUT (latency_regression > 20%):
    RECOMMENDATION = "ADOPT with optimization track" OR "DEFER to optimization phase"
ELIF (primary_metric_improvement < threshold) OR (integration_effort_high):
    RECOMMENDATION = "MONITOR" OR "REJECT"
ELSE:
    RECOMMENDATION = "REQUEST STAKEHOLDER INPUT"
```

**Report Template:**

```markdown
## Research Report: [Tool Name] vs. Current [X]

### Executive Summary
- **Recommendation:** ADOPT / REJECT / MONITOR
- **Primary Metric:** [Current] vs. [Alternative] = [Δ%] improvement
- **Secondary Metrics:** [List trade-offs]
- **Implementation effort:** [hours/days]
- **Risk:** [low/medium/high]

### Findings

#### Accuracy
- [Current tool]: [metric value]
- [Alternative tool]: [metric value]
- **Winner:** [Tool] (+X% improvement)

#### Performance (Latency)
- [Current tool]: [msec/sec audio]
- [Alternative tool]: [msec/sec audio]
- **Trade-off:** [Assessment]

#### Integration Difficulty
- **Setup:** [hours]
- **API changes:** [yes/no, describe]
- **Deployment:** [local/remote/hybrid]

### Recommendation
[Full reasoning]

### Next Steps
- [ ] Assign owner for implementation
- [ ] Open tracking issue for adoption
- [ ] Schedule integration review
```

---

## Success Criteria Template (per research area)

| Research Area | Benchmark | Primary Metric | Success Threshold | Secondary Metrics |
|---|---|---|---|---|
| **Pitch** | MIR-1K | Octave error rate | < 5% | Latency ≤ 500ms/5s audio (CPU) |
| **Stems** | MUSDB18 | SDR improvement | > 1 dB | Latency increase < 20% |
| **Chords** | Isophonics | Chord accuracy | ≥ 70% | Boundary accuracy ±50ms |
| **Lyrics** | Common Voice | WER parity | ≤ 1% increase | Latency < 0.5× WhisperX |
| **Difficulty** | Player sessions | Engagement retention | ≥ 10% improvement | Stability (lower variance) |

---

## How to Use This Guide

1. **Pick a research area** from the 5 linked issues above
2. **Follow Phase 1–4 in order** (don't skip; each builds on the last)
3. **Document findings** using the templates above
4. **Comment on the issue** with Phase completion status and key findings as you go
5. **Final report:** Post a comprehensive summary in Phase 4 section
6. **Close with decision:** Researcher + maintainer agree on ADOPT / REJECT / MONITOR

---

## Questions? Blockers?

- Unclear methodology? Comment with your question.
- Need additional benchmark datasets? Link in a comment.
- Tool installation failed? Document the error and we'll troubleshoot together.

**This task is considered complete when all 5 research areas have Phase 4 reports and adoption decisions.**
